File: sorting-searching/sorting-algorithms/Spread-sort/spread-sort.py
def spread_sort(arr, bucket_count):
    min_value = min(arr)
    max_value = max(arr)

    value_range = max_value - min_value + 1

    buckets = [[] for _ in range(bucket_count)]

    for num in arr:
        bucket_index = (
            (num - min_value)
            * bucket_count
            // value_range
        )

        if bucket_index == bucket_count:
            bucket_index -= 1

        buckets[bucket_index].append(num)

    result = []

    for bucket in buckets:
        bucket.sort()

        result.extend(bucket)

    return result


arr = [
    42, 7, 99, 15,
    63, 28, 84
]

print("Original array:")
print(arr)

sorted_array = spread_sort(arr, 4)

print("\nSorted array:")
print(sorted_array)

File: sorting-searching/sorting-algorithms/Spread-sort/spread-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> spreadSort(vector<int>& arr, int bucketCount) {
    int minValue = *min_element(arr.begin(), arr.end());
    int maxValue = *max_element(arr.begin(), arr.end());

    int range = maxValue - minValue + 1;

    vector<vector<int>> buckets(bucketCount);

    for (int num : arr) {
        int bucketIndex =
            (num - minValue) * bucketCount / range;

        if (bucketIndex == bucketCount) {
            bucketIndex--;
        }

        buckets[bucketIndex].push_back(num);
    }

    vector<int> result;

    for (auto& bucket : buckets) {
        sort(bucket.begin(), bucket.end());

        result.insert(
            result.end(),
            bucket.begin(),
            bucket.end()
        );
    }

    return result;
}

void printArray(const vector<int>& arr) {
    for (int value : arr) {
        cout << value << " ";
    }

    cout << endl;
}

int main() {
    vector<int> arr = {
        42, 7, 99, 15,
        63, 28, 84
    };

    cout << "Original array:\n";
    printArray(arr);

    vector<int> sorted =
        spreadSort(arr, 4);

    cout << "\nSorted array:\n";
    printArray(sorted);

    return 0;
}
